src/Streamly/Data/Serialize/Instances/ByteString.hs

```diff
@@ -0,0 +1,84 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE UnboxedTuples #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+module Streamly.Data.Serialize.Instances.ByteString () where
+
+import Streamly.Internal.Data.Serialize (Serialize(..))
+import Data.ByteString.Internal as Strict
+import Data.ByteString.Lazy as Lazy
+import qualified Streamly.Internal.Data.Unbox as Unbox
+import GHC.Exts
+import Foreign.ForeignPtr (withForeignPtr)
+import Foreign.Ptr (plusPtr)
+import GHC.Base (IO(..))
+import Control.Monad (foldM_)
+
+--------------------------------------------------------------------------------
+-- Strict ByteString
+--------------------------------------------------------------------------------
+instance Serialize Strict.ByteString where
+    {-# INLINE size #-}
+    size i (Strict.PS _ _ l) = i + l + size 0 l
+
+    {-# INLINE deserialize #-}
+    deserialize off arr end = do
+        (off1, lenBytes) <- deserialize off arr end :: IO (Int, Int)
+        fp <- Strict.mallocByteString lenBytes
+        let arrS# = Unbox.getMutableByteArray# arr
+            !(I# srcStartBytes#)    = off1
+            !(I# lenBytes#)         = lenBytes
+        withForeignPtr fp $ \(Ptr addr#) -> IO $ \s# -> (# copyMutableByteArrayToAddr#
+                                                        arrS# srcStartBytes# addr# lenBytes# s#
+                                                        , () #)
+        return (off1 + lenBytes, Strict.PS fp 0 lenBytes)
+
+    {-# INLINE serialize #-}
+    serialize off arr (Strict.PS fp srcOffset lenBytes) = do
+        off1 <- serialize off arr lenBytes
+        let arrD# = Unbox.getMutableByteArray# arr
+            !(I# dstStartBytes#) = off1
+            !(I# lenBytes#) = lenBytes
+        withForeignPtr fp $ \srcPtr -> let !(Ptr addr#) = srcPtr `plusPtr` srcOffset
+                                         in IO $ \s# -> (# copyAddrToByteArray#
+                                                addr# arrD# dstStartBytes# lenBytes# s#
+                                            , () #)
+        pure (off1 + lenBytes)
+
+
+--------------------------------------------------------------------------------
+-- Lazy ByteString
+--------------------------------------------------------------------------------
+instance Serialize Lazy.ByteString where
```

Comment:
You can use the template haskell helper for deriving serialization for lazy bytestring.
```
$(Serialize.deriveSerialize ''Lazy.ByteString)
```
should work.

Although I'm not sure if it will be portable across bytestring upgrades.
Changing the name and order of constructors will change the deriving strategy.

We should have more config options in the template haskell helpers with which we can give an order to the constructors manually.

Reply:
See https://github.com/composewell/streamly/issues/2592